### PR TITLE
o/snapstate: add component type to sequence points and ComponentSetup

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Storing the component type in the state is necessary so kernel-modules
type can be properly handled by the initramfs.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
Sits on top of https://github.com/snapcore/snapd/pull/13465

**Comment by alfonsosanchezbeato** [comment]
Rebased after prereq has been merged

File: overlord/snapstate/handlers_components.go [hunk 1]
```diff
@@ -266,18 +266,19 @@ func (m *SnapManager) doLinkComponent(t *state.Task, _ *tomb.Tomb) error {
...
+	if err := snapSt.Sequence.AddComponentForRevision(snapInfo.Revision, cs); err != nil {
 		return fmt.Errorf("internal error while linking component: %w", err)
 	}
 
 	// Store last component refresh time
 	if snapSt.LastCompRefreshTime == nil {
 		snapSt.LastCompRefreshTime = make(map[string]time.Time)
 	}
-	snapSt.LastCompRefreshTime[csi.Component.ComponentName] = timeNow()
+	snapSt.LastCompRefreshTime[compSetup.CompSideInfo.Component.ComponentName] = timeNow()
```
Comment (pedronis) [inline overlord/snapstate/handlers_components.go:281]: seems maybe ComponentSetup should grow a ComponentName method?

File: overlord/snapstate/sequence/sequence.go [hunk 2]
```diff
@@ -30,19 +30,30 @@ import (
...
+	Components []*ComponentState
 }
 
 // revisionSideInfoMarshal is an ancillary structure used exclusively to
 // help marshaling of RevisionSideInfo.
 type revisionSideInfoMarshal struct {
 	// SideInfo is included for compatibility with older snapd state files
 	*snap.SideInfo
-	CompSideInfo []*snap.ComponentSideInfo `yaml:"components,omitempty" json:"components,omitempty"`
+	CompSideInfo []*ComponentState `json:"components,omitempty"`
```
Comment (pedronis) [inline overlord/snapstate/sequence/sequence.go:56]: should we change the struct field name here? I don't have a storng opinion myself

File: overlord/snapstate/snapmgr.go
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:1]: the changed inner loops here maybe should become methods on SnapSequence? the first two can be just one method do as the difference is struct vs bool that can be done here?

**Review by pedronis — COMMENTED** [review]
thanks, did a pass

File: overlord/snapstate/handlers_components.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/handlers_components.go]: Added

File: overlord/snapstate/sequence/sequence.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/sequence/sequence.go]: Changed it now, this was actually an overlook

File: overlord/snapstate/snapmgr.go
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go:1]: Good point, I have created these methods now

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks, I've addressed your comments now

**Review by MiguelPires — APPROVED** [review]
lgtm

File: overlord/snapstate/snapmgr.go [hunk 3]
```diff
@@ -204,6 +204,11 @@ func NewComponentSetup(csi *snap.ComponentSideInfo, compType snap.ComponentType,
 	}
 }
 
+// ComponentName returns the component name for compsu
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:207]: missing . at the end

File: overlord/snapstate/handlers_components.go [hunk 4]
```diff
@@ -278,7 +278,7 @@ func (m *SnapManager) doLinkComponent(t *state.Task, _ *tomb.Tomb) error {
 	if snapSt.LastCompRefreshTime == nil {
 		snapSt.LastCompRefreshTime = make(map[string]time.Time)
 	}
-	snapSt.LastCompRefreshTime[compSetup.CompSideInfo.Component.ComponentName] = timeNow()
+	snapSt.LastCompRefreshTime[compSetup.ComponentName()] = timeNow()
```
Comment (pedronis) [inline overlord/snapstate/handlers_components.go:281]: is there anywhere else this could be used?

File: overlord/snapstate/sequence/sequence.go [hunk 5]
```diff
@@ -180,3 +180,28 @@ func (snapSeq *SnapSequence) RemoveComponentForRevision(snapRev snap.Revision, c
 	// might be nil
 	return unlinkedComp
 }
+
+// ComponentSideInfoForRev returns cref's component side info for the revision
+// (sequence point) indicated by revIdx if there is one.
+func (snapSeq *SnapSequence) ComponentSideInfoForRev(revIdx int, cref naming.ComponentRef) *snap.ComponentSideInfo {
```
Comment (pedronis) [inline overlord/snapstate/sequence/sequence.go:186]: we might want to produce a nicer panic if revIdx is out of bound?

**Review by pedronis — APPROVED** [review]
thanks, small notes

File: overlord/snapstate/handlers_components.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/handlers_components.go:281]: I found a couple more places, the issue is that in some cases we store CompSideInfo and use the name from there so usage gets a bit hidden.

File: overlord/snapstate/sequence/sequence.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/sequence/sequence.go:186]: Not sure if I get this one, do you mean the panic in `CurrentComponentSideInfo`? or that this method should panic?

Comment (pedronis) [reply overlord/snapstate/sequence/sequence.go:186]: just wondering if in the case if we have bug, hopefully caught by tests, were the revIdx is out of bound, wheher producing a panic here with some more context is useful before or if  the runtime one from [revIdx ]is good enough?

**Pull request merged**